iceoryx_examples/automotive_soa/include/owl/kom/event_subscriber.inl

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_SUBSCRIBER_INL
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_SUBSCRIBER_INL
+
+#include "owl/kom/event_subscriber.hpp"
+
+namespace owl
+{
+namespace kom
+{
+template <typename T>
+inline EventSubscriber<T>::EventSubscriber(const core::String& service,
+                                           const core::String& instance,
+                                           const core::String& event) noexcept
+    : m_subscriber({service, instance, event},
+                   {QUEUE_CAPACITY, HISTORY_REQUEST, iox::NodeName_t(), NOT_OFFERED_ON_CREATE})
+{
+}
+
+template <typename T>
+inline void EventSubscriber<T>::Subscribe(std::size_t) noexcept
+{
+    // maxSampleCount is ignored, because it is an argument to the c'tor of m_subscriber as part of SubscriberOptions.
+    // Utilizing late initalization by wrapping m_subscriber in an cxx::optional and calling the c'tor here would be an
+    // option
+    m_subscriber.subscribe();
+}
+
+template <typename T>
+inline void EventSubscriber<T>::Unsubscribe() noexcept
+{
+    m_subscriber.unsubscribe();
+}
+
+template <typename T>
+template <typename Callable>
+inline core::Result<size_t> EventSubscriber<T>::GetNewSamples(Callable&& callable, size_t maxNumberOfSamples) noexcept
+{
+    IOX_DISCARD_RESULT(maxNumberOfSamples);
+
+    core::Result<size_t> numberOfSamples{0};
+
+    while (m_subscriber.take()
+               .and_then([&](const auto& sample) {
+                   callable(sample.get());
+                   numberOfSamples++;
+               })
+               .or_else([](auto& result) {
+                   if (result != iox::popo::ChunkReceiveResult::NO_CHUNK_AVAILABLE)
+                   {
+                       std::cerr << "Error receiving chunk!" << std::endl;
+                   }
+               }))
+    {
+    }
+    return numberOfSamples;
+}
+
+//! [EventSubscriber setReceiveHandler]
+template <typename T>
+inline void EventSubscriber<T>::SetReceiveHandler(EventReceiveHandler handler) noexcept
+{
+    std::lock_guard<iox::posix::mutex> guard(m_mutex);
+    m_listener
```

Comment:
The listener is completely threadsafe and does not require a mutex guard.

Reply:
Removed.

Reply:
Wouldn't it be better to put the `m_receiveHandler` into a `smart_lock`. Otherwise I have no clue what the mutex named `m_mutex` should guard and it is error prone as well since one can forget to lock it.
Hint: One can use also a recursive lock with the `smart_lock`

Reply:
Of course you're right :grin: Changed.